Sources/FirebaseFirestore/FirestoreErrorCode.swift

```diff
@@ -19,33 +40,42 @@ public struct FirestoreErrorCode: Error {
+  public static let deadlineExceeded: Self = .init(firebase.firestore.kErrorDeadlineExceeded)
+  public static let notFound: Self = .init(firebase.firestore.kErrorNotFound)
+  public static let alreadyExists: Self = .init(firebase.firestore.kErrorAlreadyExists)
+  public static let permissionDenied: Self = .init(firebase.firestore.kErrorPermissionDenied)
+  public static let resourceExhausted: Self = .init(firebase.firestore.kErrorResourceExhausted)
+  public static let failedPrecondition: Self = .init(firebase.firestore.kErrorFailedPrecondition)
+  public static let aborted: Self = .init(firebase.firestore.kErrorAborted)
+  public static let outOfRange: Self = .init(firebase.firestore.kErrorOutOfRange)
+  public static let unimplemented: Self = .init(firebase.firestore.kErrorUnimplemented)
+  public static let `internal`: Self = .init(firebase.firestore.kErrorInternal)
+  public static let unavailable: Self = .init(firebase.firestore.kErrorUnavailable)
+  public static let dataLoss: Self = .init(firebase.firestore.kErrorDataLoss)
+  public static let unauthenticated: Self = .init(firebase.firestore.kErrorUnauthenticated)
 }
 
 extension FirestoreErrorCode: Equatable {}
 
-// Unfortunately `Error` is not defined as a `enum class` so we need to add
-// these wrappers.
-extension FirestoreErrorCode.Code {
-  public static var ok: Self { firebase.firestore.kErrorOk }
-  public static var none: Self { firebase.firestore.kErrorNone }
-  public static var cancelled: Self { firebase.firestore.kErrorCancelled }
-  public static var unknown: Self { firebase.firestore.kErrorUnknown }
-  public static var invalidArgument: Self { firebase.firestore.kErrorInvalidArgument }
-  public static var deadlineExceeded: Self { firebase.firestore.kErrorDeadlineExceeded }
-  public static var notFound: Self { firebase.firestore.kErrorNotFound }
-  public static var alreadyExists: Self { firebase.firestore.kErrorAlreadyExists }
-  public static var permissionDenied: Self { firebase.firestore.kErrorPermissionDenied }
-  public static var resourceExhausted: Self { firebase.firestore.kErrorResourceExhausted }
-  public static var failedPrecondition: Self { firebase.firestore.kErrorFailedPrecondition }
-  public static var aborted: Self { firebase.firestore.kErrorAborted }
-  public static var outOfRange: Self { firebase.firestore.kErrorOutOfRange }
-  public static var unimplemented: Self { firebase.firestore.kErrorUnimplemented }
-  public static var `internal`: Self { firebase.firestore.kErrorInternal }
-  public static var unavailable: Self { firebase.firestore.kErrorUnavailable }
-  public static var dataLoss: Self { firebase.firestore.kErrorDataLoss }
-  public static var unauthenticated: Self { firebase.firestore.kErrorUnauthenticated }
+extension FirestoreErrorCode {
+  public typealias Code = FirestoreErrorCode
+
+  // This allows us to re-expose self as a code similarly
+  // to what the Firebase SDK does when it creates the
+  // underlying NSErrors on iOS/macOS.
```

Comment:
Can we remove the comment and instead try to use `NSError` as a base type?

Reply:
I've tried, and it didn't work out. Why do you think it should extend from `NSError`?

Reply:
It unfortunately doesn't work because we need to support code like this:

```
switch (error as? AuthErrorCode)?.code {
case .networkError, .operationNotAllowed, .userDisabled, .wrongPassword, .invalidEmail, .tooManyRequests:
    ...
```

`NSError.code` is an `Int` :-(

Reply:
I think that a `CaseIterable` conformance should let you do this even though `NSError.code` is not an `Int`.

Reply:
> Why do you think it should extend from NSError?

~~~
// This allows us to re-expose self as a code similarly
// to what the Firebase SDK does when it creates the
// underlying NSErrors on iOS/macOS
~~~

The way that I read this is that you want the behaviour of `NSError` on Darwin on non-Darwin platforms. If that isn't the case, the comment is confusing for me. If we want the behaviour of `NSError`, then it seems to me that we should be a `NSError`.

Reply:
> I think that a CaseIterable conformance should let you do this even though NSError.code is not an Int.

I'm not sure I understand what you are proposing. Maybe you could sketch it out in more detail?

What I did do was play around with defining a `~=` operator,  but one of the problems I then hit was that some error values are duplicated on both `AuthErrorCode` and `FirestoreErrorCode`. In particular both define `.cancelled`. If I allow Swift to consider matching `Int` against `{Auth,Firestore}ErrorCode`, then it gets confused about which `.cancelled` I'm referencing. I have to then qualify that, which is not required on macOS.

Honestly, there's a lot of magic happening around `NSError` it seems on Apple's platforms, and I haven't quite wrapped my head around it.

The comment you are referencing was here already before I started hacking on this code. I think we don't actually need to implement `NSError` to make Arc happy. I could be wrong, but so far that doesn't appear to be necessary.

So my plan is to not try to force-fit `NSError` right now (because it doesn't seem to work), and revisit if necessary. Already this PR is a huge improvement enabling much more code sharing on the Arc side.

Reply:
The `CaseIterable` conformance allows you to switch over the type.  Something like this should do the trick:

```
extension AuthErrorCode: CaseIterable {
  public static var allCases: [Int] {
    return [1, 2, 3]
  }
}
```

This will allow the structure to behave as an enumerated type.

Reply:
but wouldn't this run into the same issue with `.cancelled` being an ambiguous reference to either `AuthErrorCode.cancelled` or `FirestoreErrorCode.cancelled`?

Reply:
`AuthErrorCode` and `FirestoreErrorCode` are two separate types, so there shouldn't be any ambiguity. I feel like we might be talking past each other - how does the type ambiguity arise if I am saying that the only possible values are explicitly stated and that the values are from two separate enumerations. We could very easily have an inner `enum` that would give us the type safety and ability to use that to disambiguate.

```swift
extension AuthErrorCode {
  public enum Value {
    case there_can_be_only_one
    case there_is_only_zuul
  }
}

extension AuthErrorCode: CaseIterable {
  public static var allCases: [AuthErrorCode.Value] {
    [.there_can_be_only_one, .there_is_only_zuul]
  }
}
```

Reply:
As discussed over Tuple, we are going to postpone extending from `NSError`. That is something we can do in a follow-up if / when it becomes valuable.